### PR TITLE
ESEF 2022 validations

### PR DESCRIPTION
**Pull request by gmongelli**
#### Reason for change
A new version of the ESEF Reporting Manual has been published on 5 August 2022.

The current ESEF validations need to be updated accordingly.

In this pull request, some preliminary improvements are integrated. 

#### Description of change
The current ESEF validation is left untouched, but a new validation plug-in called 'ESEF-2022' has been added.

It currently includes the following changes:
- **1.3.3**: Detect empty text block elements?
- **1.6.1**: Verify that when the balance attribute is present the amounts are positive (they may only be negative if the balance attribute is not set). **_Please review carefully this change to see if it is appropriate!_**
- **2.2.4**: Detect multi-language duplicates (String only) within a single iXBRL report.
- **2.2.6**: Validate if “escape” attribute is present in tag introducing a text block if the text contains a “<” or “&” character
- **2.5.1**: Return a validation error if the iXBRL report contains embedded mailto links.
- **3.2.1**: Remove any validation rules related to the naming conventions for extension taxonomy elements
- **3.2.5**: Remove the validation rule “Extension taxonomy SHOULD NOT define abstract concepts.”
- **3.3.1**: Check if extension taxonomy elements are anchored to abstract concepts. This should not happen.
- **3.4.5**: Check that for any element (both core and extension) at most one label is defined for a given role and language
- **3.4.6**: Implement the validation rule “All usable concepts in extension taxonomy relationships SHOULD be applied by tagged facts. “ 

#### Steps to Test
A sample instance for taxonomy version 2021 will be posted soon.

**review**:
@Arelle/arelle

### CODE REVIEW

**Comment by aviary2-wf** [comment]
## Security Insights
No security relevant content was detected by automated scans.

## Action Items
* Review PR for [security impact](https://wiki.atl.workiva.net/display/SECURITY/Development+Security+Review+Guidelines); comment "security review required" if needed or unsure
* Verify [`aviary.yaml`](https://wiki.atl.workiva.net/pages/viewpage.action?pageId=42762752#SecurityInsightFramework&Tooling-ConfiguringAviary) coverage of security relevant code
---
_Questions or Comments? Reach out on Slack: #support-infosec._

File: arelle/plugin/validate/ESEF/__init__.py
Comment (derekgengenbacher-wf) [inline arelle/plugin/validate/ESEF/__init__.py]: This check should also probably contain an xValid check so that we do not run into weird instances where xValue is a string.

File: arelle/plugin/validate/ESEF/DTS.py
Comment (derekgengenbacher-wf) [inline arelle/plugin/validate/ESEF/DTS.py]: Can we just remove this from the code as well, please?

File: arelle/plugin/validate/ESEF-2022/Util.py
Comment (austinmatherne-wk) [inline arelle/plugin/validate/ESEF-2022/Util.py:149]: Initialize the `_aspectEqualFacts` dict inside of the `_factConceptContextUnitHash.values()` loop so you don't need to call `_aspectEqualFacts.clear()` at the end:
```suggestion
    for hashEquivalentFacts in _factConceptContextUnitHash.values():
        _aspectEqualFacts = defaultdict(dict)
        for f in hashEquivalentFacts:  # check for hash collision by value checks on context and unit
```

**Review by austinmatherne-wk — COMMENTED** [review]
Really good work. Thank you for opening this!

File: arelle/plugin/validate/ESEF/__init__.py
Comment (gmongelli) [reply arelle/plugin/validate/ESEF/__init__.py]: The suggested check on xValid has already been done some lines higher:

https://github.com/Arelle/Arelle/blob/50a5692f74a2f4c64ae1186714bc3f7e762009fe/arelle/plugin/validate/ESEF/__init__.py#L774

File: arelle/plugin/validate/ESEF/DTS.py
Comment (gmongelli) [reply arelle/plugin/validate/ESEF/DTS.py]: Requested change will be pushed soon.

File: arelle/plugin/validate/ESEF/DTS.py
Comment (gmongelli) [reply arelle/plugin/validate/ESEF/DTS.py]: Requested change will be pushed soon.

File: arelle/plugin/validate/ESEF-2022/Util.py
Comment (gmongelli) [reply arelle/plugin/validate/ESEF-2022/Util.py:168]: Change will be pushed soon.

File: arelle/plugin/validate/ESEF-2022/Util.py
Comment (gmongelli) [reply arelle/plugin/validate/ESEF-2022/Util.py:159]: I am not sure if your suggestion is really equivalent. I let @geoffreyPiraprez decide...

**Comment by ggravlingen** [comment]
@gmongelli looks like you are typing this strictly, in line with the current module, which is great! Can you please add the package to pyproject.toml so that it gets picked up by the CI/CD as typed?

File: arelle/plugin/validate/ESEF_2022/Dimensions.py
Comment (derekgengenbacher-wf) [inline arelle/plugin/validate/ESEF_2022/Dimensions.py:5]: nit: typo in `Manual`

File: arelle/plugin/validate/ESEF_2022/__init__.py
Comment (derekgengenbacher-wf) [inline arelle/plugin/validate/ESEF_2022/__init__.py:7]: Just a comment, no work needed. Its really unfortunate that ESMA overwrote the older url with the new data. It would have been nice to be able to keep the documentation on the specs separate.

On a side note does anyone know if a new RTS has been issued to accompany the new guidelines?

File: arelle/plugin/validate/ESEF-2022/Util.py
Comment (austinmatherne-wk) [reply arelle/plugin/validate/ESEF-2022/Util.py]: Initially I thought facts with None units were filtered out, but that's not the case. None contexts should be though.

**Review by austinmatherne-wk — DISMISSED** [review]

File: arelle/plugin/validate/ESEF_2022/__init__.py
Comment (gmongelli) [reply arelle/plugin/validate/ESEF_2022/__init__.py:7]: Hello,

Here is the consolidated version that puts together all rules on that matter:

https://eur-lex.europa.eu/legal-content/EN/TXT/?uri=CELEX%3A02019R0815-20220101

Maybe we can put that version in the file comments...

Greg

File: arelle/plugin/validate/ESEF_2022/Dimensions.py
Comment (gmongelli) [reply arelle/plugin/validate/ESEF_2022/Dimensions.py:5]: I have changed the whole name and added the same reference as in arelle/plugin/validate/ESEF_2022/__init__.py

The changes will be pushed soon

File: arelle/plugin/validate/ESEF_2022/__init__.py
Comment (austinmatherne-wk) [inline arelle/plugin/validate/ESEF_2022/__init__.py:713]: Required for type checking/CI (this is in the other ESEF plugin):
```suggestion
                modelObject=contextsWithPeriodTime, contextIds=", ".join(c.id for c in contextsWithPeriodTime if c.id))
```

File: arelle/plugin/validate/ESEF_2022/__init__.py
Comment (austinmatherne-wk) [inline arelle/plugin/validate/ESEF_2022/__init__.py:717]: Required for type checking/CI (this is in the other ESEF plugin):
```suggestion
                modelObject=contextsWithPeriodTimeZone, contextIds=", ".join(c.id for c in contextsWithPeriodTimeZone if c.id))
```

File: arelle/plugin/validate/ESEF_2022/Util.py
Comment (austinmatherne-wk) [inline arelle/plugin/validate/ESEF_2022/Util.py:47]: Required for type checking/CI (this is in the other ESEF plugin).
```suggestion
        return False
    assert modelObject.qname is not None
    ns = modelObject.qname.namespaceURI
    assert ns is not None
    return (any(ns.startswith(esefNsPrefix) for esefNsPrefix in esefTaxonomyNamespaceURIs))
```

File: arelle/plugin/validate/ESEF_2022/Util.py
Comment (austinmatherne-wk) [inline arelle/plugin/validate/ESEF_2022/Util.py:134]: Required for type checking:
```suggestion
def checkForMultiLangDuplicates(modelXbrl: ModelXbrl) -> None:

    _factConceptContextUnitHash: defaultdict[int, list[ModelInlineFact]] = defaultdict(list)
```

File: arelle/plugin/validate/ESEF_2022/Util.py
Comment (austinmatherne-wk) [inline arelle/plugin/validate/ESEF_2022/Util.py:149]: Required for type checking (and erroneous semicolon):
```suggestion
            continue
        _aspectEqualFacts: defaultdict[tuple[QName, str], dict[tuple[ModelContext, ModelUnit | None], list[ModelInlineFact]]] = defaultdict(dict)
```

**Review by austinmatherne-wk — COMMENTED** [review]
Looks good other than the failing mypy type checking. We can merge this once the type checking errors are cleared and CI passes.

File: arelle/plugin/validate/ESEF_2022/Const.py
Comment (sagesmith-wf) [inline arelle/plugin/validate/ESEF_2022/Const.py:21]: Nit: Applies to every multiline call in this fine, indent is wrong for all of these

Comment (sagesmith-wf) [reply arelle/plugin/validate/ESEF_2022/Const.py:21]: This looks like a duplicate entry and perhaps missing the year?

File: arelle/plugin/validate/ESEF_2022/Const.py
Comment (sagesmith-wf) [inline arelle/plugin/validate/ESEF_2022/Const.py:54]: I would suggest only duplicating the files that you needed to change between versions. The constant files seem fine to share

Comment (austinmatherne-wk) [reply arelle/plugin/validate/ESEF_2022/Const.py:54]: Sharing code between plugins is murky. Right now the priority is to get this merged into the master branch of Arelle. Once it's in we'll work on folding it into the existing ESEF plugin.

File: arelle/plugin/validate/ESEF_2022/Const.py
Comment (austinmatherne-wk) [reply arelle/plugin/validate/ESEF_2022/Const.py:21]: Preexisting: we're really only looking at making updates in this PR for rule changes in 2022. We'll work on cleaning up issues with the existing plugin.

File: pyproject.toml
Comment (austinmatherne-wk) [inline pyproject.toml:106]: Hi @gmongelli. If you remove this line we can merge this PR. We're happy to add type hints and get type checking working for ESEF 2022 after this initial PR is merged.

**Review by austinmatherne-wk — APPROVED** [review]

**Review by derekgengenbacher-wf — APPROVED** [review]
Validation changes look solid.

**Pull request merged**
